src/util/string_utils.cpp

```diff
@@ -87,7 +106,11 @@ void split_string(
   std::string &right,
   bool strip)
 {
-  assert(!std::isspace(delim));
+  if(strip && std::isspace(delim))
+  {
+    throw std::invalid_argument(
+      "delim can't be a space character if using strip");
+  }
```

Comment:
Is there a reason that this condition/throw block can't just be an INVARIANT ?

Reply:
For the unit test

Reply:
I am with @chrisr-diffblue: I think this should be a `PRECONDITION`. I think `invariant.h` ought to have the facilities to turn that into something that can be caught, and hence could be used for testing error paths in unit tests.

Reply:
I don't agree - I think of all the code we write, `util` code is the most "library like" and not throwing limits the utility of this function. Imagine Using `split_string` on a user input: I now have to duplicate the error handling code if I want to allow the user to try again.

Reply:
I think @thk123 does make a good point about the 'library like' use-case - perhaps the answer would be to have 'CATCHABLE_PRECONDITION' along with @tautschnig suggestion to enable error path testing? - i.e. in the library use-case Thomas is talking about, it could be caught by the caller to avoid duplicating the error handling.

Reply:
There are two different "users" to be distinguished: the developer using the library and the end-user feeding input to what the developer has built. The end-user should not be subjected to internal limitations, but for a developer it's reasonable to assume that they have read the documentation.

It is of course perfectly reasonable to use exceptions for either class of "users," but that's not what the code base does right now. If you think this should change, I'd go via a new issue and/or have some further discussion. Because your proposed code is of course very nice and clean I'd keep it so that a future patch to introduce the exceptions is very low effort.

Reply:
In all this I do think that `PRECONDITION_STRUCTURED` should be able to support this. There are a few cases of `INVARIANT_STRUCTURED` in the code base. (But also `invariant.h` says something about the future...)

Reply:
Sorry I'm not clear on what you propose I do with this code for this PR? Do you mean keep the code in this PR, or change this PR to a precondition (and remove the associated unit tests) and stash this version away for a later day?

Reply:
Here is what I'd do if it were my PR: if things can be made to work via `PRECONDITION_STRUCTURED`, use it. Else stash it away for a later day and drop a note in an issue so that it gets tracked.